File: python/questions/minCost1.py
from typing import List

class Solution:
    def minCost(self, grid: List[List[int]], k: int) -> int:
        r = len(grid)
        c = len(grid[0])
        maxn = -float('inf')
        for row in grid:
            maxn = max(maxn, max(row))
        gmj = [float('inf') for _ in range(maxn+2)]
        cmj = [float('inf') for _ in range(maxn+1)]
        for j in range(k+1):
            rowc = [float('inf') for _ in range(c+1)]
            rowc[1] = -grid[0][0]
            for ri in range(r):
                for ci in range(c):
                    cv = grid[ri][ci]
                    rowc[ci+1] = min(min(rowc[ci], rowc[ci+1])+cv, gmj[cv])
                    cmj[cv] = min(cmj[cv], rowc[ci+1])
            for n in range(maxn, -1, -1):
                gmj[n] = min(gmj[n+1], cmj[n])
        return rowc[-1]

File: cpp/questions/minCost1.cpp
#include <vector>
#include <climits>
#include <algorithm>

using namespace std;

class Solution {
public:
    int minCost(vector<vector<int>>& grid, int k) {
        int r = grid.size();
        int c = grid[0].size();
        int maxn = INT_MIN;
        for (auto& row: grid) {
            maxn = std::max(maxn, *max_element(row.begin(), row.end()));
        }
        std::vector<int> rowc(c+1, INT_MAX);               
        std::vector<int> gmj(maxn+2, INT_MAX);
        std::vector<int> cmj(maxn+1, INT_MAX);
        for (int j=0; j<=k; j++) {
            std::fill(rowc.begin(), rowc.end(), INT_MAX);
            rowc[1] = -grid[0][0];
            for (int ri=0; ri<r; ri++) {
                for (int ci=0; ci<c; ci++) {
                    int cv = grid[ri][ci];
                    rowc[ci+1] = std::min(std::min(rowc[ci], rowc[ci+1])+cv, gmj[cv]);
                    cmj[cv] = std::min(cmj[cv], rowc[ci+1]);
                }
            }
            for (int js=maxn; js>=0; js--) {
                gmj[js] = std::min(gmj[js+1], cmj[js]);
            }
        }
        return rowc[c];
    }
};
